### PR TITLE
feat: add data download api

### PR DESCRIPTION
**Pull request by hansol-medi**
Implementation of data download api returning the data with zip format.

In this PR, the data is processed through pipelined channel(referred [go.dev blog](https://go.dev/blog/pipelines)). 
For every round, `svc.setDataPipeline` sets channel for pipelining data. 
The channels are added to `merger` struct, and the `merger` merges the channels and write data using zipWriter. 
There is a channel for error as well. When error occurs in downloading data, it sends signal to the channel, and close all the channels.

To serve raw data to redeemer, the handler do the below things.
- get data certs from panacea
- download encrypted data from the storage
- decrypt and return it

Currently, it is supported only in the form of zip of a byte array, but it can be possible to expand it to csv or other format in the future.

FYI) there are some temp functions, vars and a handler for make and get sample certs & data. It will be removed when adding verification of redeem receipt.

##

### Updates

The `merger` struct is no longer needed for data pipelining. It was an over-engineering for now, so removed.

When getting data, a goroutine is created by round. It's important to handle multiple goroutines when both error occurs as well as  writing data. So I used `errgroup` to handle errors correctly in multiple goroutines.

And the zip writer is not goroutine-safe to write data concurrently in multiple goroutines. To write the data to zip writer safely, made a new struct, `ConcurrentZipWriter`, which has mutex to enable lock and unlock when writing data.

### CODE REVIEW

File: server/service/datapool/download_data_handler.go [hunk 1]
```diff
@@ -1,10 +1,139 @@
...
+	// decrypt data
+	data, err := crypto.DecryptDataWithAES256(svc.DataEncKey, nil, cipherData)
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func validateStreamRequest(r *http.Request) (error, int) {
```
Comment (gyuguen) [inline server/service/datapool/download_data_handler.go:132]: Actually we are responding with a Content-Type of `application/zip`.
It seems that the content type to be verified needs to be modified.

**Review by gyuguen — COMMENTED** [review]
Thank you to your hard work.
I left a few comments.
And let's do some more actual behavior tests.

File: server/service/datapool/download_data_handler.go [hunk 2]
```diff
@@ -1,10 +1,139 @@
...
+
+	out := make(chan []byte, len(certs))
+
+	go func() {
+		defer close(out)
+
+		for _, cert := range certs {
+			select {
+			case <-errPipeline:
+				return
```
[the remark this answers is not in the document]
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go:90]: When I tested, if I wait the error from `errPipeline` chan as you suggested, it caused deadlock.
So just detected the signal from `errPipeline` and return, resulting in calling deferred close of `out` channel.
I will test again. Thanks!

File: server/service/datapool/download_data_handler.go (hunk 1 above)
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go:132]: Ok. I thought the user do not know which content type is valid, so just accept `application/octet-stream` and set response header `application/zip`.
Or we can accept both content type, `application/zip` and `application/octet-stream`

File: server/service/datapool/download_data_handler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go:90]: No it does not cause deadlock. With the code you suggested, however, all the go routines don't stop when error occur. Let me take a look at it more.

[the remark this answers is not in the document]
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go:90]: This select case statement is for receiving err from `errPipeline` chan. 

> My guess is that if you get an error here on the error channel, you won't get it on the parent function to error channel.

@gyuguen, may I ask what's the meaning of not getting error on the parent function? When error occur, all the go routines in `setDataPipeline()` and `Merger.Merge()` receive the error from the `errPipeline`. When go routine detects error, they return and call deferred function of closing channel or wg.Done.

File: server/service/datapool/download_data_handler.go (hunk 1 above)
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go]: Changed to accept `application/zip` for download request in [2560226](https://github.com/medibloc/panacea-data-market-validator/pull/75/commits/256022608ab0b88d02d72d066f6c915e89454155)

**Review by youngjoon-lee — COMMENTED** [review]
Please don't merge this PR before I review. After a quick look, I've found many things to be improved.

**Comment by hansol-medi** [comment]
> Please don't merge this PR before I review. After a quick look, I've found many things to be improved.

Ok. It will be helpful!

**Comment by hansol-medi** [comment]
As @gyuguen said ([comment](https://github.com/medibloc/panacea-data-market-validator/pull/75#discussion_r869047219)), when multiple errors occur, the error channel waits infinitely. 
So I changed to use `errgroup` to handle error(s) in multiple go routines. With these changes, the description of this PR becomes quite different. After we discuss about this PR, let's make this PR updated or create a new one.

**Comment by hansol-medi** [comment]
Thanks to @youngjoon-lee's nice advice, I could further improve the download api with goroutine-safety.
Please review once again and feel free to leave your comments!

Got a few todos left, I'll do this for next PR. The below is the list.

TODO
- remove all the temp codes regarding sample data
  - add getting redeem receipt from panacea
  - add verification of redeem receipt (then replace temp `poolID` and `redeemedAtRound`)
- remove temp api to store sample data (`server/service/datapool/sample_data_handler.go`)
- file format of data 
  - It could be `json` or `csv`, and so on. How do we set the format of the data? query string of request?
- add test code by mocking storage

File: server/service/datapool/download_data_handler.go [hunk 3]
```diff
@@ -1,10 +1,115 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func validateDownloadRequest(r *http.Request) (error, int) {
+	if r.Header.Get("Content-Type") != "application/zip" {
+		return fmt.Errorf("only application/zip is supported"), http.StatusUnsupportedMediaType
```
[the remark this answers is not in the document]
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go]: In validation of request, checking content type is removed. Thanks!

File: server/service/datapool/download_data_handler.go [hunk 4]
```diff
@@ -1,10 +1,115 @@
...
+		g.Go(func() error {
+			for _, cert := range certs {
+				data, err := svc.downloadAndDecryptData(cert)
+				if err != nil {
+					return err
+				}
+
+				if _, err := zipWriter.Write(data); err != nil {
+					return err
+				}
```
[the remark this answers is not in the document]
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go]: True. It is modified to be goroutine-safe as you advised.

File: server/service/datapool/download_data_handler.go [hunk 5]
```diff
@@ -1,10 +1,115 @@
...
+	poolIDTemp := uint64(1)
+	redeemedRoundTemp := uint64(3)
+
+	filename := "pool-" + strconv.FormatUint(poolIDTemp, 10) + "-data"
+
+	zw := zip.NewWriter(w)
+
+	zipWriter, err := zw.Create(filename)
+	if err != nil {
+		log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go]: replaced it to use htte.Error, with log.Error()

File: server/service/datapool/download_data_handler.go [hunk 6]
```diff
@@ -1,10 +1,115 @@
...
+
+	// TODO: get poolID and round from redeem receipt. For now, temp value
+	poolIDTemp := uint64(1)
+	redeemedRoundTemp := uint64(3)
+
+	filename := "pool-" + strconv.FormatUint(poolIDTemp, 10) + "-data"
+
+	zw := zip.NewWriter(w)
+
+	zipWriter, err := zw.Create(filename)
```
[the remark this answers is not in the document]
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go]: As you mentioned, each file is generated for every data.

File: server/service/datapool/download_data_handler.go [hunk 7]
```diff
@@ -1,10 +1,115 @@
...
+				if _, err := zipWriter.Write(data); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
```
[the remark this answers is not in the document]
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go:77]: You're right! it is fixed in [256fcc6](https://github.com/medibloc/panacea-data-market-validator/pull/75/commits/256fcc6ff5ef5b1061240d0022d4db21539a23c8). When error occurs, all the goroutines would be stopped by `<-ctx.Done()`.

File: server/service/datapool/download_data_handler.go [hunk 8]
```diff
@@ -1,10 +1,124 @@
...
+	}
+
+	w.Header().Set("Content-Type", "application/zip")
+	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"pool-%d.zip\"", poolIDTemp))
+
+	if err := czw.Close(); err != nil {
+		log.Errorf("error occurred while closing zip writer: %v", err)
+		return
+	}
+
```
Comment (inchori) [inline server/service/datapool/download_data_handler.go:90]: Does we also need to add `http.Errorf` in this part?

**Review by inchori — COMMENTED** [review]
Thanks for your effort!. I leave one question.

**Comment by hansol-medi** [comment]
> It would be good to be able to add tests to types/concurrent_zip_writer.go.

A simple test for `concurrent_zip_writer.go` has been added in [052fd94](https://github.com/medibloc/panacea-data-market-validator/pull/75/commits/052fd94fa80bce55e2aa0254a8f93336a03fb905). 
Please review again :)

**Comment by gyuguen** [comment]
> > It would be good to be able to add tests to types/concurrent_zip_writer.go.
> 
> A simple test for `concurrent_zip_writer.go` has been added in [052fd94](https://github.com/medibloc/panacea-data-market-validator/pull/75/commits/052fd94fa80bce55e2aa0254a8f93336a03fb905). Please review again :)

Good.
This test looks good.

File: server/service/datapool/download_data_handler.go [hunk 9]
```diff
@@ -1,10 +1,129 @@
...
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		log.Errorf("failed to download: %v", err)
+		http.Error(w, "failed to download", http.StatusInternalServerError)
+		return
+	}
+
+	if err := czw.Close(); err != nil {
```
[the remark this answers is not in the document]
Comment (hansol-medi) [reply server/service/datapool/download_data_handler.go:87]: You're right. I have to close the `czw` in defer. Thanks!

File: server/service/datapool/sample_data_handler.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/medibloc/panacea-data-market-validator/crypto"
+	log "github.com/sirupsen/logrus"
+)
+
+// handleSampleData is a temporary handler to make sample data
+func (svc *dataPoolService) handleSampleData(w http.ResponseWriter, r *http.Request) {
```
Comment (inchori) [inline server/service/datapool/sample_data_handler.go:15]: I have a question. Will the `handleSampleData` method be removed soon?

**Review by youngjoon-lee — APPROVED** [review]
lgtm. thank you for trying this pattern

**Pull request merged**